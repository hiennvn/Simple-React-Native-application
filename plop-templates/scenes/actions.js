const DEFAULT_ACTION = 'DEFAULT_ACTION'

export const actions = {
  DEFAULT_ACTION,
}

export const defaultAction = () => {
  return async dispatch => {
    dispatch({
      type: DEFAULT_ACTION,
      value: [],
    })
  }
}
