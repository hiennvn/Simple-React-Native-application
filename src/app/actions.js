const APP_TRACKER = 'APP_TRACKER'
const APP_USER_TOKEN = 'APP_USER_TOKEN'
const APP_USER_INFO = 'APP_USER_INFO'
const APP_STORE = 'APP_STORE'
const APP_REALM = 'APP_REALM'

export const AppActions = {
    APP_TRACKER,
    APP_USER_TOKEN,
    APP_USER_INFO,
    APP_STORE,
    APP_REALM
}