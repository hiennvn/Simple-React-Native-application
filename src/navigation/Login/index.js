import { StackNavigator } from 'react-navigation'

import Login from '@scenes/Login'
import Onboarding from '@scenes/Onboarding'
import { Routes } from '../routes'

/**
 * This is a nested navigator, so we just pass it's config
 */
export default StackNavigator({
  [Routes.Login]: { screen: Login },
  [Routes.Onboarding]: { screen: Onboarding },
})
