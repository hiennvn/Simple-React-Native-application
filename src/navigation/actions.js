export default {
  MOVE: 'MOVE',
  RESET: 'RESET',
}

// It is not working with jest tests
// Works live. TODO check if React navigation fixed it:
// https://github.com/react-community/react-navigation/issues/2820
// export { Routes as LoginRoutes } from './Login'
// export { Routes as MainRoutes } from './Main'
// It is not working with jest tests
