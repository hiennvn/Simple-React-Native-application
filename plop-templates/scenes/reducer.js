import { actions } from './actions'

const initialState = {}

export default (state = initialState, action) => {
  switch (action.type) {
    case actions.DEFAULT_ACTION:
      return Object.assign({}, state, {
        items: action.value,
      })
    default:
      return state
  }
}
