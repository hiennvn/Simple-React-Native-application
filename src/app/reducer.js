import { getGATracker } from '../services/google-analytics'
import { AppActions } from './actions'
import ServiceFactory from '../services/factory';

const initialState = {
    tracker: getGATracker(),
    accessToken: null,
    userInfo: null,
    realm: null,
    serviceFactory: null
  }

export default (state = initialState, action) => {
  switch(action.type) {
    case AppActions.APP_TRACKER:
      return Object.assign({}, state, {
        tracker: action.value
      });
    case AppActions.APP_USER_TOKEN:
      return Object.assign({}, state, {
        accessToken: action.value
      });
    case AppActions.APP_USER_INFO:
      return Object.assign({}, state, {
        userInfo: action.value
      });
    case AppActions.APP_STORE:
      if(action.tracker) {
        return Object.assign({}, state, {
          tracker: action.tracker,
          accessToken: action.accessToken,
          userInfo: action.userInfo
        });
      } else {
        return Object.assign({}, state, {
          accessToken: action.accessToken,
          userInfo: action.userInfo
        });
      }
    case AppActions.APP_REALM:
      return Object.assign({}, state, {
        realm: action.value,
        serviceFactory: new ServiceFactory(action.value)
      });
    default:
      return state;
  }
}