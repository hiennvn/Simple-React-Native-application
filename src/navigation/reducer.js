import { NavigationActions } from 'react-navigation'

import { MenuNavigator as Navigator } from './App'
import actions from './actions'

const initialState = Navigator.router.getStateForAction(
  NavigationActions.init()
)

export default (state = initialState, action) => {
  switch (action.type) {
    case actions.MOVE:
      return Navigator.router.getStateForAction(
        NavigationActions.navigate({ routeName: action.value }),
        state
      )
    case actions.RESET:
      return Navigator.router.getStateForAction(
        NavigationActions.reset({
          index: 0,
          actions: [NavigationActions.navigate({ routeName: action.value })],
        })
      )
    default:
      return Navigator.router.getStateForAction(action, state)
  }
}
