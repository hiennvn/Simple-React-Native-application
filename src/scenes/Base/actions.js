import { Routes } from '../../navigation/routes'
import { AppActions } from '@app/actions'
import NavigatorActions from '@navigation/actions'
import UserService from '../../services/biz/user'

export const updateTracker = (tracker) => {
    return {
      type: AppActions.APP_TRACKER,
      value: tracker,
    }
  }
  
  export const setAccessToken = (accessToken) => {
    return {
      type: AppActions.APP_USER_TOKEN,
      value: accessToken,
    }
  }
  
  export const setUserInfo = (userInfo) => {
    return {
      type: AppActions.APP_USER_INFO,
      value: userInfo,
    }
  }

  export const setAppStore = (tracker, accessToken, userInfo) => {
    return {
      type: AppActions.APP_STORE,
      tracker: tracker,
      accessToken: accessToken,
      userInfo: userInfo
    }
  }

  export const setRealm = (realm) => {
    return {
      type: AppActions.APP_REALM,
      value: realm,
    }
  }
  
  export const moveToMain = () => {
    return {
      type: NavigatorActions.MOVE,
      value: Routes.Main,
    }
  }
  
  export const moveToLogin = () => {
    return {
      type: NavigatorActions.MOVE,
      value: Routes.Login,
    }
  }

  export const getShortUserNameTitle = (appStore) => {
    let userService = appStore.serviceFactory.getService(UserService.name);

    return userService.getShortUserNameTitle(appStore.userInfo);
  }
  
  export const getUserName = (appStore) => {
    let userService = appStore.serviceFactory.getService(UserService.name);

    return userService.getUserName(appStore.userInfo);
  }
  
  export const getUserProfilePicture = (appStore) => {
    let userService = appStore.serviceFactory.getService(UserService.name);

    return userService.getUserProfilePicture(appStore.userInfo);
  }

  export const getUserCoverPicture = (appStore) => {
    let userService = appStore.serviceFactory.getService(UserService.name);

    return userService.getUserCoverPicture(appStore.userInfo);
  }