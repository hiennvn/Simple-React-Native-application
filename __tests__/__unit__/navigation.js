import 'react-native'
import React from 'react'
import configureStore from 'redux-mock-store'
import { shallow } from 'enzyme'

import { LoginNavigator, Routes as LoginRoutes } from '@navigation/Login'
import Navigator, {
  MainNavigator,
  Routes as MainRoutes,
} from '@navigation/App'

import reducer from '@navigation/reducer'
import actions from '@navigation/actions'

const state = {
  navigation: {
    index: 0,
    routes: [{ routeName: MainRoutes.Splash }],
  },
}
const mockStore = configureStore()

describe('Navigators', () => {
  it('render correctly', () => {
    const wrapper = shallow(<Navigator store={mockStore(state)} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('are configured correctly', () => {
    expect(LoginNavigator).toMatchSnapshot()
    expect(MainNavigator).toMatchSnapshot()
  })
})

describe('Navigation reducer', () => {
  it('reduces initial state', () => {
    const initialState = reducer(undefined, { type: undefined })
    expect(initialState.index).toBe(0)
    expect(initialState.routes[0].routeName).toBe(MainRoutes.Splash)
  })

  it('reduces first depth nested navigation', () => {
    Object.values(MainRoutes).map(route => {
      const newState = reducer(undefined, {
        type: actions.MOVE,
        value: route,
      })
      expect(newState.index).toBe(1)
      expect(newState.routes[newState.index].routeName).toBe(route)
    })
  })

  it('reduces second depth nested navigation', () => {
    Object.values(MainRoutes).map(route => {
      const newState = reducer(
        reducer(undefined, {
          type: actions.MOVE,
          value: MainRoutes.Main,
        }),
        {
          type: actions.MOVE,
          value: route,
        }
      )
      expect(newState.index).toBe(2)
      expect(newState.routes[newState.index].routeName).toBe(route)
    })
  })
})
