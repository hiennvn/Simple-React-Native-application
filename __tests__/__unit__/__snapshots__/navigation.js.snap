// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigators are configured correctly 1`] = `undefined`;

exports[`Navigators are configured correctly 2`] = `[Function]`;

exports[`Navigators render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Navigator)
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "function",
    "props": Object {
      "dispatch": [Function],
      "state": Object {
        "navigation": Object {
          "index": 0,
          "routes": Array [
            Object {
              "routeName": "Splash",
            },
          ],
        },
      },
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "function",
      "props": Object {
        "dispatch": [Function],
        "state": Object {
          "navigation": Object {
            "index": 0,
            "routes": Array [
              Object {
                "routeName": "Splash",
              },
            ],
          },
        },
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
